File: lintcode/binary-tree-level-order-traversal.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/binary-tree-level-order-traversal/

"""
Definition of TreeNode:
"""
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None

class Solution:
    """
    @param root: The root of binary tree.
    @return: Level order in a list of lists of integers
    """
    def levelOrder(self, root):
        # write your code here
        if not root:
            return []

        result = [[]]
        queue = [root]
        current_level_count = 1
        next_level_count = 0
        while queue:
            node = queue.pop(0)
            current_level_count -= 1
            result[-1].append(node.val)
            if node.left:
                queue.append(node.left)
                next_level_count += 1
            if node.right:
                queue.append(node.right)
                next_level_count += 1
            if current_level_count == 0:
                if next_level_count > 0:
                    result.append([])
                current_level_count, next_level_count = next_level_count, current_level_count
        return result

if __name__ == '__main__':
    s = Solution()
    root = TreeNode(3)
    root.left = TreeNode(9)
    root.right = TreeNode(20)
    root.right.left = TreeNode(15)
    root.right.right = TreeNode(7)
    print s.levelOrder(root)

File: leetcode/binary-tree-level-order-traversal.cc
// Question: https://leetcode.com/problems/binary-tree-level-order-traversal/

#include "headers.h"

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> ans;
        if (!root) {
            return ans;
        }

        queue<TreeNode*> q;
        q.push(root);
        int curr_level_count = 1;
        int next_level_count = 0;
        vector<int> level_vec;
        while (!q.empty()) {
            TreeNode* node = q.front();
            q.pop();
            level_vec.push_back(node->val);

            if (node->left) {
                q.push(node->left);
                ++next_level_count;
            }
            if (node->right) {
                q.push(node->right);
                ++next_level_count;
            }

            if (--curr_level_count == 0) {
                curr_level_count = next_level_count;
                next_level_count = 0;
                ans.push_back(level_vec);
                level_vec.clear();
            }
        }

        return ans;
    }
};

int main(int argc, char** argv) {
    Solution s;
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->left->left = new TreeNode(3);
    root->left->right = new TreeNode(4);
    vector<vector<int>> ans = s.levelOrder(root);
    for (auto v : ans) {
        for (auto i : v) {
            cout << i << " ";
        }
        cout << endl;
    }
    return 0;
}
